modules/hardware/starfive/default.nix

```diff
@@ -0,0 +1,50 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib)
+    mkIf
+    mkMerge
+    mkOption
+    types
+  ;
+  inherit (config.hardware) mmcBootIndex;
+  cfg = config.hardware.socs;
+  starFiveSOCs = [ "starfive-jh7100" ];
+  anyStarFive = lib.any (soc: config.hardware.socs.${soc}.enable) starFiveSOCs;
+  isPhoneUX = config.Tow-Boot.phone-ux.enable;
+in
+{
+  options = {
+    hardware.socs = {
+      starfive-jh7100.enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Enable when SoC is StarFive JH7100";
+        internal = true;
+      };
+    };
+  };
+
+  config = mkMerge [
+    {
+      hardware.socList = starFiveSOCs;
+    }
+    (mkIf anyStarFive {
+      Tow-Boot = {
+        system.system = "riscv64-linux";
+        # uhh...
+      };
+    })
+
+    # Documentation fragments
+    (mkIf (anyStarFive && !isPhoneUX) {
+      documentation.sections.installationInstructions =
+        lib.mkDefault
+        (config.documentation.helpers.genericInstallationInstructionsTemplate {
+          # StarFive will prefer SD card always.
```

Comment:
I don't know the platform. Is this true?

Reply:
For VisionFive, yes, I believe it always prefers SD cards.

Reply:
Specifically, u-boot will always prefer (require) an sd card for booting.
I'm not entirely sure if in the firmware , that's a hard requirement.

Reply:
Specifically, u-boot will always prefer (require) an sd card for booting.
I'm not entirely sure if in the firmware , that's a hard requirement.

Reply:
Right, I understand the confusion.

Generally speaking, there are two "orders" when you power on a system.

The *Platform Firmware startup order*, and the *operating system boot order*. I make it a point to reserve the word "boot" to operating systems only.

This comment is about the *Platform Firmware startup order*, sometimes called *BootROM order*. This is the locations the SoC will look for U-Boot, and start it from. I don't think the BootROM even looks at other locations than SPI Flash [and serial], but I have literally no proof about that. So this comment will not hold true.

Note that the *operating system boot order* is entirely controlled by the platform firmware (U-Boot, Tow-Boot), and with Tow-Boot we set it to what we want, which is "internal targets first".

Reply:
I'm not entirely sure now, if we can replace the boot stuff in the SPI, we might be able to boot from USB.

Reply:
I am not 100% sure, but given what I've read yesterday, I'm about 90% sure that the firmware will strictly be loaded from SPI Flash, and that the OS will be loaded *however we want*.

And that indeed means that if USB support is complete enough in U-Boot for the board, that it will be possible to boot from a good old boring and generic USB UEFI iso.